### ISSUE TITLE
"too much recursion" in browserified version

### ISSUE THREAD

**Issue by kibtidec** [codeschool/sqlite-parser#38]
I am struggling with the browserified version. When including `sqlite-parser.js` (taken from the `dist` directory) through the same local server as the html-file, I get an error message with `Internal error: too much recursion` (Firefox) or `RangeError: Maximum call stack size exceeded`(Chrome)
If I do include the script externally (e.g. from `rawgit.com`) then it works.

The html I am using is the following:
```{html}
<html>
        <head>
                <script src="sqlite-parser.js"></script>
        </head>
        <body>
                <textarea id="in">SELECT a FROM b</textarea>
                <button onclick="parse()">Parse</button>
                <p/>
                <div id="out"></div>
                <script>
function parse() {
query=document.getElementById("in").value;
sqliteParser(query, function (err, ast) {
  if (err) {
      console.error(err);
      return;
    }
  document.getElementById('out').innerHTML = '<pre>'+ 
     JSON.stringify(ast, null, 2) + '</pre>';
});}
                </script>
        </body>
</html>
```
if I load the **same** `sqlite-parser.js` externally, e.g. through `rawgit.com`, then it works flawlessly and the string is parsed and the AST printed.

```{html}
       <script src="https://rawgit.com/codeschool/sqlite-parser/1599a58dd08bbe55c4d14e1f6844ad13b7c9c421/dist/sqlite-parser.js"></script>
```

If I go back to older versions of `sqlite-parser.js` then the newest which works is the version from commit 85596844f45833a526f7c390b5683f1e3ed08e44 (2016-03-09). The commit after this is not available on github, and the one after this throws above errors.

**Comment by nwronski** [comment]
@kibtidec This one is pretty fascinating, and I am still not exactly sure why it is happening, but hopefully this helps a bit...

Using your test code I get the same error when serving `index.html` from my filesystem (e.g.: `file:///Users/nicholas/.../index.html`) but it works perfectly fine when serving the same file from a local server (e.g.: `localhost:3000/index.html`). I am not sure if there is some kind of stack size limit for scripts served from the filesystem versus from HTTP/HTTPS or if there is some other cause for this difference.

If you are just looking to get something set up locally to start playing around with, you can clone this repository and spin up the demo code:

````
git clone https://github.com/codeschool/sqlite-parser.git
cd sqlite-parser
npm install
grunt live
```

and then open up `http://localhost:8080` in your browser.

**Comment by kibtidec** [comment]
@nwronski thank you for your comments and hints on how to run locally ... the thing is what I really want to do is, to call the parser from within `R` through its `V8` package. Importing `sqlite-parser.js` there and calling the parser crashes my `R` session, so I tried to replicate this in Javascript in browser to figure out whether the problem is in `R` or generally.

Using from within `R` the version from commit 85596844f45833a526f7c390b5683f1e3ed08e44 works but later versions crash.

One more comment on serving to the browser locally versa remotely: I also received the recursion error message when serving the same file through webspace at my provider, whereas it worked when served through `rawgit.com` - I could not see any differences in the way it is served though (e.g. HTTP headers).

**Comment by nwronski** [comment]
@kibtidec I figured it out (kind of)! The error is introduced after uglifying the bundle. To create `dist/sqlite-parser.js` I _was_ browserifying and then uglifying the source files. Testing things out locally now, the parser works fine when running the browserified bundle but then I encounter that error when using the uglified (minified) bundle.

Until I can figure out what exactly is causing the problem, I am publishing sqlite-parser with the un-minified bundle at `dist/sqlite-parser.js`. Please try installing the latest release ([v1.0.0-rc3](https://github.com/codeschool/sqlite-parser/releases/tag/v1.0.0-rc3)) and let me know if that fixes the problem for you, thanks!

```
npm i sqlite-parser@beta
```

**Comment by kibtidec** [comment]
@nwronski Thank you. Did further tests:

In the browser:
- replacing `sqlite-parser.js` with the newest version results in the expected behaviour
- I can now run the sync version of `sqliteParser` (which previously did not work for me either)

In `V8` engine of `R` it _does not work_
- async: `ReferenceError: setTimeout is not defined`  (as `setTimeout` is not part of `V8` implementation)
- sync: `TypeError: Object function Object() { [native code] } has no method 'assign'` 

The sync version would be more natural to use in my case, but I'd be happy if either would work with the `V8` engine ;-)

**Comment by nwronski** [comment]
@kibtidec I am not super-familiar with your use case, but AFAIK `Object.assign()` is [available in the latest versions of V8](https://chromium.googlesource.com/v8/v8.git/+/fda20efb2f75a11b711501c3d8bb19cef4873200).

Are you able to use Node.js instead of V8? You'd get a `setTimeout()` implementation if you could switch to Node.

You might also try [Babel CLI](https://babeljs.io/docs/usage/cli/) to compile `sqlite-parser.js` to native ES5 if you cannot upgrade to the latest V8.

**Comment by kibtidec** [comment]
@nwronski Thank you for your comments.

In the environment I use I need to use V8 as there is an interface to V8 from within R.
I will need to dig deeper into the `Object.assign`. Debian and ubuntu use rather old  version for `libv8` and with newer versions from other repositories I was not yet able to compile the V8 library for R.

I will also check further the Babel route you were suggesting - did not know this tool yet. Maybe it is a way out. I am not too much into Javascript programming and other Node.js libraries worked well with the V8 interface.

**Comment by nwronski** [comment]
Feel free to reopen this issue if your problem is not resolved. There were some major changes released in [v1.0.0](https://github.com/codeschool/sqlite-parser/releases/tag/v1.0.0) so it might be worth trying again using the latest release.
